File: packages/mobile_ios/App/TerminalRender.swift
import CoreGraphics
import Foundation
import UIKit

// Pure presentation rules for the terminal grid (unit tested). Grid contents are
// user content: nothing here logs, and nothing is exposed to accessibility.

struct GridSize: Equatable {
    var cols: UInt16
    var rows: UInt16
}

/// Monospace cell size in points and the text baseline offset within a cell.
struct CellMetrics: Equatable {
    var width: CGFloat
    var height: CGFloat
    var baseline: CGFloat

    /// Cell width is rounded to device pixels so background runs never leave seams.
    static func monospaced(size: CGFloat, scale: CGFloat) -> CellMetrics {
        let font = UIFont.monospacedSystemFont(ofSize: max(size, 1), weight: .regular)
        let ctFont = font as CTFont
        var character: UniChar = 0x4d // "M"
        var glyph: CGGlyph = 0
        var advance = CGSize.zero
        if CTFontGetGlyphsForCharacters(ctFont, &character, &glyph, 1) {
            CTFontGetAdvancesForGlyphs(ctFont, .horizontal, &glyph, &advance, 1)
        }
        // Monospace advance is ~0.6 em; implausible measurements use that instead.
        let natural = advance.width >= size * 0.3 ? advance.width : size * 0.6
        let pixels = max(scale, 1)
        let width = max((natural * pixels).rounded() / pixels, 1 / pixels)
        let height = ceil(font.ascender - font.descender + max(font.leading, 0))
        return CellMetrics(width: width, height: height > 0 ? height : ceil(size * 1.2),
                           baseline: font.ascender > 0 ? font.ascender : size)
    }
}

let MIN_FONT_POINTS: CGFloat = 8
let MAX_FONT_POINTS: CGFloat = 32
let MAX_GRID_SIDE = 512
let MAX_GRID_CELLS = 65_536

/// Largest grid that fits, bounded by the core's limits (512 per side, 65,536 cells).
func terminalGrid(width: CGFloat, height: CGFloat, metrics: CellMetrics) -> GridSize {
    var cols = min(max(Int((width / max(metrics.width, 1)).rounded(.down)), 1), MAX_GRID_SIDE)
    let rows = min(max(Int((height / max(metrics.height, 1)).rounded(.down)), 1), MAX_GRID_SIDE)
    if cols * rows > MAX_GRID_CELLS { cols = MAX_GRID_CELLS / rows }
    return GridSize(cols: UInt16(cols), rows: UInt16(rows))
}

/// `#RRGGBB` → 0xRRGGBB; malformed values fall back to `fallback`.
func parseColor(_ value: String, _ fallback: UInt32) -> UInt32 {
    guard value.utf8.count == 7, value.hasPrefix("#"), let rgb = UInt32(value.dropFirst(), radix: 16) else { return fallback }
    return rgb
}

func cgColor(_ rgb: UInt32, alpha: CGFloat = 1) -> CGColor {
    CGColor(srgbRed: CGFloat((rgb >> 16) & 0xff) / 255, green: CGFloat((rgb >> 8) & 0xff) / 255,
            blue: CGFloat(rgb & 0xff) / 255, alpha: alpha)
}

struct TextStyle: Equatable {
    var fg: UInt32
    var bold = false
    var italic = false
    var underline = false
    var strikethrough = false
    /// Decorations are drawn even over blank cells.
    var decorated: Bool { underline || strikethrough }
}

/// `ascii` runs hold single-cell printable ASCII (blanks as spaces) drawn glyph by
/// glyph at exact cell positions; other runs are one cell (wide = 2) each.
struct TextRun: Equatable {
    var row: Int
    var col: Int
    var text: String
    var cells: Int
    var style: TextStyle
    var ascii: Bool
}

struct BgRun: Equatable {
    var row: Int
    var col: Int
    var cells: Int
    var color: UInt32
}

struct CursorBox: Equatable {
    var row: Int
    var col: Int
    var cells: Int
    var shape: TerminalCursorShape
    var color: UInt32
    /// Glyph under a block cursor, redrawn in `textColor`.
    var text: String
    var textColor: UInt32
    var style: TextStyle
}

struct RenderPlan: Equatable {
    var cols: Int
    var rows: Int
    var background: UInt32
    var backgrounds: [BgRun]
    var texts: [TextRun]
    var cursor: CursorBox?
}

private func printableASCII(_ text: String) -> Bool {
    let scalars = text.unicodeScalars
    return scalars.count == 1 && (0x21...0x7e).contains(scalars.first!.value)
}

/// Draw operations for one snapshot. Colors are already resolved by the VT (including
/// reverse-video mode); only the per-cell inverse flag is applied here. The fill is the
/// most common background; other backgrounds merge into horizontal runs.
func renderPlan(_ snapshot: TerminalSnapshot, fallbackFg: UInt32 = 0xffffff, fallbackBg: UInt32 = 0x000000) -> RenderPlan {
    let cols = Int(snapshot.cols), rows = Int(snapshot.rows)
    let cells = snapshot.cells
    let count = min(cells.count, cols * rows)
    var fgs = [UInt32](repeating: fallbackFg, count: count)
    var bgs = [UInt32](repeating: fallbackBg, count: count)
    var frequency: [UInt32: Int] = [:]
    for i in 0..<count {
        var fg = parseColor(cells[i].fg, fallbackFg)
        var bg = parseColor(cells[i].bg, fallbackBg)
        if cells[i].inverse { swap(&fg, &bg) }
        fgs[i] = fg
        bgs[i] = bg
        frequency[bg, default: 0] += 1
    }
    let background = frequency.max { $0.value != $1.value ? $0.value < $1.value : $0.key > $1.key }?.key ?? fallbackBg
    var backgrounds: [BgRun] = []
    var texts: [TextRun] = []
    func style(_ i: Int) -> TextStyle {
        TextStyle(fg: fgs[i], bold: cells[i].bold, italic: cells[i].italic,
                  underline: cells[i].underline, strikethrough: cells[i].strikethrough)
    }
    for row in 0..<rows {
        var col = 0
        while col < cols, row * cols + col < count {
            let bg = bgs[row * cols + col]
            var end = col + 1
            while end < cols, row * cols + end < count, bgs[row * cols + end] == bg { end += 1 }
            if bg != background { backgrounds.append(BgRun(row: row, col: col, cells: end - col, color: bg)) }
            col = end
        }
        var run: TextRun?
        func flush() {
            guard var current = run else { return }
            run = nil
            if !current.style.decorated {
                // Blanks only matter under decorations.
                let leading = current.text.prefix { $0 == " " }.count
                current.text = String(current.text.dropFirst(leading))
                current.col += leading
                while current.text.last == " " { current.text.removeLast() }
                current.cells = current.text.count
                if current.text.isEmpty { return }
            }
            texts.append(current)
        }
        for c in 0..<cols {
            let i = row * cols + c
            guard i < count else { break }
            let cell = cells[i]
            if cell.width == 0 { flush(); continue }
            let blank = cell.text.isEmpty || cell.text == " "
            let cellStyle = style(i)
            if cell.width == 1, blank || printableASCII(cell.text) {
                let glyph = blank ? " " : cell.text
                if run != nil, run!.style == cellStyle, run!.col + run!.cells == c {
                    run!.text += glyph
                    run!.cells += 1
                } else {
                    flush()
                    run = TextRun(row: row, col: c, text: glyph, cells: 1, style: cellStyle, ascii: true)
                }
                continue
            }
            flush()
            texts.append(TextRun(row: row, col: c, text: cell.text, cells: cell.width == 2 ? 2 : 1, style: cellStyle, ascii: false))
        }
        flush()
    }
    var cursor: CursorBox?
    let at = snapshot.cursor
    if at.visible, snapshot.scroll_offset == 0, Int(at.row) < rows, Int(at.col) < cols {
        let i = Int(at.row) * cols + Int(at.col)
        let inside = i < count
        cursor = CursorBox(row: Int(at.row), col: Int(at.col), cells: inside && cells[i].width == 2 ? 2 : 1,
                           shape: at.shape, color: inside ? fgs[i] : fallbackFg,
                           text: inside && cells[i].text != " " ? cells[i].text : "",
                           textColor: inside ? bgs[i] : background,
                           style: inside ? style(i) : TextStyle(fg: fallbackFg))
    }
    return RenderPlan(cols: cols, rows: rows, background: background, backgrounds: backgrounds, texts: texts, cursor: cursor)
}

/// A linear (stream) selection between two grid cells, inclusive, in either drag direction.
struct GridSelection: Equatable {
    var anchorRow: Int
    var anchorCol: Int
    var row: Int
    var col: Int

    private var ordered: ((Int, Int), (Int, Int)) {
        (anchorRow, anchorCol) <= (row, col) ? ((anchorRow, anchorCol), (row, col)) : ((row, col), (anchorRow, anchorCol))
    }
    var start: (row: Int, col: Int) { ordered.0 }
    var end: (row: Int, col: Int) { ordered.1 }
    func contains(_ r: Int, _ c: Int) -> Bool { start <= (r, c) && (r, c) <= end }
}

/// Selected text; trailing blanks per line are trimmed, and lines join with `\n`.
func selectedText(_ snapshot: TerminalSnapshot, _ selection: GridSelection) -> String {
    let cols = Int(snapshot.cols)
    let (startRow, startCol) = selection.start
    let (endRow, endCol) = selection.end
    guard cols > 0, startRow <= min(endRow, Int(snapshot.rows) - 1) else { return "" }
    var lines: [String] = []
    for row in max(startRow, 0)...min(endRow, Int(snapshot.rows) - 1) {
        let from = row == startRow ? startCol : 0
        let to = row == endRow ? min(endCol, cols - 1) : cols - 1
        var line = ""
        if from <= to {
            for col in from...to {
                let index = row * cols + col
                guard index < snapshot.cells.count else { break }
                let cell = snapshot.cells[index]
                if cell.width == 0 { continue }
                line += cell.text.isEmpty ? " " : cell.text
            }
        }
        while line.last == " " { line.removeLast() }
        lines.append(line)
    }
    return lines.joined(separator: "\n")
}

File: packages/mobile_android/app/src/main/java/dev/verdeai/app/TerminalRender.kt
package dev.verdeai.app

import dev.verdeai.core.TerminalCursorShape
import dev.verdeai.core.TerminalSnapshot
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

// Pure presentation rules for the terminal canvas (unit tested). Grid contents are
// user content: nothing here logs, and nothing is exposed to semantics.

/** Monospace cell size in pixels and the text baseline offset within a cell. */
internal data class CellMetrics(val width: Float, val height: Float, val baseline: Float)

/** Largest grid that fits, bounded by the core's limits (512 per side, 65,536 cells). */
internal fun terminalGrid(widthPx: Float, heightPx: Float, metrics: CellMetrics): Pair<Int, Int> {
    var cols = floor(widthPx / max(metrics.width, 1f)).toInt().coerceIn(1, MAX_GRID_SIDE)
    val rows = floor(heightPx / max(metrics.height, 1f)).toInt().coerceIn(1, MAX_GRID_SIDE)
    if (cols * rows > MAX_GRID_CELLS) cols = MAX_GRID_CELLS / rows
    return cols to rows
}

internal const val MAX_GRID_SIDE = 512
internal const val MAX_GRID_CELLS = 65_536

/** `#RRGGBB` → opaque ARGB; malformed values fall back to [fallback]. */
internal fun parseColor(value: String, fallback: Int): Int =
    if (value.length == 7 && value[0] == '#') value.substring(1).toIntOrNull(16)?.let { it or (0xff shl 24) } ?: fallback
    else fallback

internal data class TextRun(val row: Int, val col: Int, val text: String, val cells: Int,
    val fg: Int, val bold: Boolean, val italic: Boolean, val underline: Boolean, val strikethrough: Boolean)
internal data class BgRun(val row: Int, val col: Int, val cells: Int, val color: Int)
internal data class CursorBox(val row: Int, val col: Int, val cells: Int, val shape: TerminalCursorShape, val color: Int, val textColor: Int)
internal data class RenderPlan(val cols: Int, val rows: Int, val background: Int, val backgrounds: List<BgRun>,
    val texts: List<TextRun>, val cursor: CursorBox?)

/**
 * Draw operations for one snapshot. Colors are already resolved by the VT (including
 * reverse-video mode); only the per-cell inverse flag is applied here. The canvas fill is
 * the most common background, and other backgrounds merge into horizontal runs.
 */
internal fun renderPlan(snapshot: TerminalSnapshot, fallbackFg: Int = 0xffffffff.toInt(), fallbackBg: Int = 0xff000000.toInt()): RenderPlan {
    val cols = snapshot.cols
    val rows = snapshot.rows
    val cells = snapshot.cells
    val count = min(cells.size, cols * rows)
    val fgs = IntArray(count)
    val bgs = IntArray(count)
    val frequency = HashMap<Int, Int>()
    for (i in 0 until count) {
        val cell = cells[i]
        var fg = parseColor(cell.fg, fallbackFg)
        var bg = parseColor(cell.bg, fallbackBg)
        if (cell.inverse) { val swap = fg; fg = bg; bg = swap }
        fgs[i] = fg; bgs[i] = bg
        frequency.merge(bg, 1, Int::plus)
    }
    val background = frequency.maxByOrNull { it.value }?.key ?: fallbackBg
    val backgrounds = ArrayList<BgRun>()
    val texts = ArrayList<TextRun>()
    for (row in 0 until rows) {
        var col = 0
        while (col < cols) {
            val i = row * cols + col
            if (i >= count) break
            val bg = bgs[i]
            var end = col + 1
            while (end < cols && row * cols + end < count && bgs[row * cols + end] == bg) end++
            if (bg != background) backgrounds += BgRun(row, col, end - col, bg)
            col = end
        }
        for (c in 0 until cols) {
            val i = row * cols + c
            if (i >= count) break
            val cell = cells[i]
            if (cell.width == 0 || cell.text.isEmpty() || cell.text == " ") continue
            texts += TextRun(row, c, cell.text, if (cell.width == 2) 2 else 1, fgs[i], cell.bold, cell.italic, cell.underline, cell.strikethrough)
        }
    }
    val cursor = snapshot.cursor.takeIf { it.visible && snapshot.scroll_offset == 0L && it.row in 0 until rows && it.col in 0 until cols }?.let {
        val i = it.row * cols + it.col
        val wide = i < count && cells[i].width == 2
        CursorBox(it.row, it.col, if (wide) 2 else 1, it.shape, if (i < count) fgs[i] else fallbackFg, if (i < count) bgs[i] else background)
    }
    return RenderPlan(cols, rows, background, backgrounds, texts, cursor)
}

/** A linear (stream) selection between two grid cells, inclusive, in either drag direction. */
internal data class GridSelection(val anchorRow: Int, val anchorCol: Int, val row: Int, val col: Int) {
    private val startKey get() = min(anchorRow * 1_000 + anchorCol, row * 1_000 + col)
    private val endKey get() = max(anchorRow * 1_000 + anchorCol, row * 1_000 + col)
    fun contains(r: Int, c: Int) = (r * 1_000 + c) in startKey..endKey
    val start get() = startKey / 1_000 to startKey % 1_000
    val end get() = endKey / 1_000 to endKey % 1_000
}

/** Selected text; trailing blanks per line are trimmed, and lines join with `\n`. */
internal fun selectedText(snapshot: TerminalSnapshot, selection: GridSelection): String {
    val cols = snapshot.cols
    val (startRow, startCol) = selection.start
    val (endRow, endCol) = selection.end
    val lines = ArrayList<String>()
    for (row in startRow.coerceAtLeast(0)..min(endRow, snapshot.rows - 1)) {
        val from = if (row == startRow) startCol else 0
        val to = if (row == endRow) min(endCol, cols - 1) else cols - 1
        val line = StringBuilder()
        for (col in from..to) {
            val cell = snapshot.cells.getOrNull(row * cols + col) ?: break
            if (cell.width == 0) continue
            line.append(cell.text.ifEmpty { " " })
        }
        lines += line.trimEnd().toString()
    }
    return lines.joinToString("\n")
}
